sshd-common/src/main/java/org/apache/sshd/common/config/keys/impl/OpenSSHCertificateDecoder.java

```diff
@@ -0,0 +1,115 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.sshd.common.config.keys.OpenSshCertificate;
+import org.apache.sshd.common.keyprovider.KeyPairProvider;
+import org.apache.sshd.common.session.SessionContext;
+import org.apache.sshd.common.util.buffer.ByteArrayBuffer;
+import org.apache.sshd.common.util.buffer.keys.OpenSSHCertPublicKeyParser;
+import org.apache.sshd.common.util.io.IoUtils;
+
+/**
+ * @author <a href="mailto:dev@mina.apache.org">Apache MINA SSHD Project</a>
+ */
+public class OpenSSHCertificateDecoder extends AbstractPublicKeyEntryDecoder<OpenSshCertificate, OpenSshCertificate> {
+    public static final OpenSSHCertificateDecoder INSTANCE = new OpenSSHCertificateDecoder();
+
+    public OpenSSHCertificateDecoder() {
+        super(OpenSshCertificate.class, OpenSshCertificate.class,
+                Collections.unmodifiableList(Arrays.asList(
+                        KeyPairProvider.SSH_RSA_CERT,
+                        KeyPairProvider.SSH_DSS_CERT,
+                        KeyPairProvider.SSH_ED25519_CERT,
+                        KeyPairProvider.SSH_ECDSA_SHA2_NISTP256_CERT,
+                        KeyPairProvider.SSH_ECDSA_SHA2_NISTP384_CERT,
+                        KeyPairProvider.SSH_ECDSA_SHA2_NISTP521_CERT
+                )));
+    }
+
+    @Override
+    public OpenSshCertificate decodePublicKey(
+            SessionContext session, String keyType, InputStream keyData, Map<String, String> headers)
+            throws IOException, GeneralSecurityException {
+
+        byte[] bytes = IoUtils.toByteArray(keyData);
+
+        ByteArrayBuffer buffer = new ByteArrayBuffer(bytes);
+
+        OpenSshCertificate cert = (OpenSshCertificate) OpenSSHCertPublicKeyParser.INSTANCE.getRawPublicKey(keyType, buffer);
+
+        if (cert.getType() != OpenSshCertificate.SSH_CERT_TYPE_HOST) {
+            throw new GeneralSecurityException("The provided certificate is not a Host certificate.");
+        }
+
+        return cert;
+    }
+
+    @Override
+    public String encodePublicKey(OutputStream s, OpenSshCertificate key) throws IOException {
+        Objects.requireNonNull(key, "No public key provided");
+
+        ByteArrayBuffer buffer = new ByteArrayBuffer();
+        buffer.putRawPublicKeyBytes(key);
+        s.write(buffer.getCompactData());
+
+        return key.getKeyType();
+    }
+
+    @Override
+    public OpenSshCertificate clonePublicKey(OpenSshCertificate key) throws GeneralSecurityException {
+        try {
+            ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+            String keyType = encodePublicKey(outStream, key);
+            InputStream inStream = new ByteArrayInputStream(outStream.toByteArray());
+            return decodePublicKey(null, keyType, inStream, null);
+        } catch (IOException e) {
+            throw new GeneralSecurityException("Unable to clone key.", e);
+        }
+    }
```

Comment:
Even though we are dealing with byte-array streams that do no require closing, as a matter of coding style I prefer to have them in try-with-resource blocks:
```java
try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
    ...
    try (InputStream inStream = new ByteArrayInputStream(outStream.toByteArray())) {
         ...
         return ....;
    }
}
```